### PR TITLE
Restrict the Details window to only 1 instance

### LINKED ISSUE #24445: Restrict the `Details` window to only 1 instance
### Expected behavior
Clicking the `Details` button opens a window. Clicking it again either: closes the window, or focuses it.

### Actual behavior
Clicking the button repeatedly opens multiple windows. Running an algorithm then crashes the Workbench

### Steps to reproduce the behavior
Click the `Details` button many times. Then run an algorithm.

### Platforms affected
Windows

### PR DESCRIPTION
**Pull request by dtasev**
**Description of work.**
Allows the user to only have 1 `Details` window. Additionally fixes a memory 'leak' where the presenter's close function wouldn't be called when the `Details` window is closed from the `X` - this triggers closeEvent that previously wasn't overriden to call the presenter's method. This is done by funnelling both the `X` and clicking the `Close` button events through the view's `closeEvent`, which then calls the presenter's `close`.

**To test:**
Click `Details` a bunch of times. Only 1 window should open at all times. Try closing via both the `X` and via the `Close` button.

Fixes #24445 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqt/widgets/algorithmprogress/dialog_presenter.py [hunk 1]
```diff
@@ -13,9 +13,13 @@ class AlgorithmProgressDialogPresenter(AlgorithmProgressPresenterBase):
...
     Presents progress reports on algorithms.
     """
+
     def __init__(self, view, model):
         super(AlgorithmProgressDialogPresenter, self).__init__()
-        view.close_button.clicked.connect(self.close)
+        # connect the close button to the closeEvent of the window
+        # so that pressing the X button, and pressing `Close` go through
+        # the same routine, and properly call the presenter's close()
+        view.close_button.clicked.connect(view.closeEvent)
```
Comment (mantid-roman) [inline qt/python/mantidqt/widgets/algorithmprogress/dialog_presenter.py]: The connection should be to `view.close` which will in turn call `closeEvent(event)` passing in a proper `QCloseEvent`.

File: qt/python/mantidqt/widgets/algorithmprogress/dialog_widget.py [hunk 2]
```diff
@@ -67,6 +70,16 @@ def __init__(self, parent, model):
         self.presenter = AlgorithmProgressDialogPresenter(self, model)
         self.presenter.update_gui()
 
+    def closeEvent(self, *args):
```
Comment (mantid-roman) [inline qt/python/mantidqt/widgets/algorithmprogress/dialog_widget.py]: The argument should be `event` of type `QCloseEvent` according to the documentation http://doc.qt.io/qt-5/qwidget.html#closeEvent

File: qt/python/mantidqt/widgets/algorithmprogress/dialog_widget.py [hunk 3]
```diff
@@ -67,6 +70,16 @@ def __init__(self, parent, model):
...
+        """
+        Funnel the closeEvent, triggered when the user presses X,
+        through the same routine as the `Close` button
+        :param args:
+        :return:
+        """
+        self.presenter.close()
+        self.deleteLater()
+
```
Comment (mantid-roman) [inline qt/python/mantidqt/widgets/algorithmprogress/dialog_widget.py]: Call `event.accept()` at the end.

File: qt/python/mantidqt/widgets/algorithmprogress/dialog_widget.py [hunk 4]
```diff
@@ -38,8 +39,10 @@ class AlgorithmMonitorDialog(QDialog):
     """
     Displays progress of all running algorithms.
     """
+
     def __init__(self, parent, model):
         super(AlgorithmMonitorDialog, self).__init__(parent)
+        self.parent = parent
```
Comment (mantid-roman) [inline qt/python/mantidqt/widgets/algorithmprogress/dialog_widget.py]: There is no need to save `parent` as it is passed to the base class constructor and can be accessed via `self.parent()` method.

File: qt/python/mantidqt/widgets/algorithmprogress/dialog_presenter.py [hunk 5]
```diff
@@ -60,7 +64,7 @@ def close(self):
         """
         self.model.remove_presenter(self)
         self.progress_bars.clear()
-        self.view.close()
+        self.view.parent.clear_dialog()
```
Comment (mantid-roman) [inline qt/python/mantidqt/widgets/algorithmprogress/dialog_presenter.py]: Use `parent()`

**Review by mantid-roman — CHANGES_REQUESTED** [review]

**Comment by DTasev** [comment]
Just pushed up changes that address all the comments

**Review by mantid-roman — APPROVED** [review]

**Pull request merged**